Rubberduck.VBEditor.VB6/SafeComWrappers/VB/VBProject.cs

```diff
@@ -187,32 +186,116 @@ public string ProjectDisplayName
                     return _displayName;
                 }
 
-                var vbe = VBE;
-                var activeProject = vbe.ActiveVBProject;
-                var mainWindow = vbe.MainWindow;
+                _displayName = DisplayNameFromFileName();
+
+                if (string.IsNullOrEmpty(_displayName))
+                {
+                    _displayName = DisplayNameFromWindowCaption();
+                }
+
+                if (string.IsNullOrEmpty(_displayName)
+                    || _displayName.EndsWith("..."))
                 {
-                    try
+                    var nameFromBuildFileName = DisplayNameFromBuildFileName();
+                    if (!string.IsNullOrEmpty(nameFromBuildFileName)
+                        && nameFromBuildFileName.Length > _displayName.Length - 3) //Otherwise, we got more of the name from the previous attempt.
                     {
-                        if (Target.HelpFile != activeProject.HelpFile)
-                        {
-                            vbe.ActiveVBProject = this;
-                        }
+                        _displayName = nameFromBuildFileName;
+                    }
+                }
+
+                return _displayName;
+            }
+        }
+
+        private string DisplayNameFromFileName()
+        {
+            return Path.GetFileName(FileName);
+        }
+
+        private string DisplayNameFromBuildFileName()
+        {
+            var pseudoDllName = Path.GetFileName(BuildFileName);
+            return pseudoDllName == null || pseudoDllName.Length <= 4 //Should not happen as the string should always end in .DLL.
+                ? string.Empty
+                : pseudoDllName.Substring(0, pseudoDllName.Length - 4);
+        }
+
+        private static readonly Regex CaptionProjectRegex = new Regex(@"^(?:[^-]+)(?:\s-\s)(?<project>.+)(?:\s-\s.*)?$");
+        private static readonly Regex OpenModuleRegex = new Regex(@"^(?<project>.+)(?<module>\s-\s\[.*\((Code|UserForm)\)\])$");
+        private static readonly Regex PartialOpenModuleRegex = new Regex(@"^(?<project>.+)(\s-\s\[)");
+        private static readonly Regex NearlyOnlyProject = new Regex(@"^(?<project>.+)(\s-?\s?)$");
+
+        private string DisplayNameFromWindowCaption()
+        {
+            using (var vbe = VBE)
+            using (var activeProject = vbe.ActiveVBProject)
+            using (var mainWindow = vbe.MainWindow)
+            {
+                try
+                {
+                    if (ProjectId != activeProject.ProjectId)
+                    {
+                        vbe.ActiveVBProject = this;
+                    }
+
+                    var caption = mainWindow.Caption;
+                    if (caption.Length > 99)
+                    {
+                        //The value returned will be truncated at character 99 and the rest is garbage due to a bug in the VBE API.
```

Comment:
As discussed, project name for VB6 seems limited to 40 chars so this *probably* won't apply (except perhaps in locales where the application name is lengthy?)